### PR TITLE
Reuse registration on network error

### PR DESCRIPTION
**Pull request by chmielewskiandreas**
Based on PR https://github.com/zephyrproject-rtos/zephyr/pull/45476/commits/b662799fe96706bed8f0c6f1db23d0c7a5837815 this PR covers the situation to reuse a registration on a network error. In a later patch my goal is to offer the application more control over network errors and to be able to suspend/resume connection in non-queue-mode

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 1]
```diff
@@ -258,15 +268,19 @@ static void socket_fault_cb(int error)
...
 
-	client.ctx->sec_obj_inst = -1;
+	lwm2m_engine_context_close_keep_observers(client.ctx);
 
 	/* Jump directly to the registration phase. In case there is no valid
 	 * security object for the LWM2M server, it will fall back to the
 	 * bootstrap procedure.
 	 */
-	set_sm_state(ENGINE_DO_REGISTRATION);
+	set_sm_state(ENGINE_NETWORK_ERROR);
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:283]: Comment above is now misleading. Just remove it.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 2]
```diff
@@ -318,6 +332,8 @@ static int do_bootstrap_reply_cb(const struct coap_packet *response,
 		COAP_RESPONSE_CODE_DETAIL(code));
 
 	if (code == COAP_RESPONSE_CODE_CHANGED) {
+		/* remember the last reg time */
+		client.last_update = k_uptime_get();
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:336]: I don't think we should update the timer here. Bootstrapped is not yet registered.

File: subsys/net/lib/lwm2m/lwm2m_rd_client.c [hunk 3]
```diff
@@ -1065,7 +1086,23 @@ static void sm_do_network_error(void)
...
+	if (err) {
+		LOG_ERR("Failed to start socket %d", err);
+		/* keep this state until lifetime/retry count exceeds. Renew sm state
+		 * to set retry_delay etc ... */
+		set_sm_state(ENGINE_NETWORK_ERROR);
+		return;
+	}
+
+	engine_trigger_update(false);
+	set_sm_state(ENGINE_REGISTRATION_DONE);
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_rd_client.c:1105]: This state change worries me.
`NETWORK_ERROR ---(invalid state change)---> REGISTRATION_DONE --> UPDATE_SENT`

I see that by doing this, you are triggering the functionality in `sm_registration_done()`. But this is now hard to spot, and not obvious at all for developer looking at this code. If we need calls from `sm_registration_done()` maybe it needs to split out from there, so we can directly call the function and stay correctly in `NETWORK_ERROR` state. Or maybe we should consider going into `RX_OFF` or something like that.

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]
I would like to see the error handling done in error state without causing any side effects of transitioning the client into REGISTERED state when in fact, it is not registered.

**Comment by chmielewskiandreas** [comment]
Any update/comments on my last commit?

**Review by rlubos — COMMENTED** [review]
Looks fine to me, but let's wait for @SeppoTakalo feedback as he was mostly involved in the review.

**Comment by chmielewskiandreas** [comment]
@SeppoTakalo What do you think about the changes I did?

**Comment by chmielewskiandreas** [comment]
> @SeppoTakalo What do you think about the changes I did?

I would like to hear your opinion before I do any kind of rebase here and invest time for that. Thanks :)

**Review by SeppoTakalo — DISMISSED** [review]
I think this now looks as intended.

Can you also update the documentation and a diagram from here:
https://docs.zephyrproject.org/latest/connectivity/networking/api/lwm2m.html#lwm2m-engine-and-application-events

The diagram is in a .png format that can be opened in [draw.io](https://app.diagrams.net/) then add the add the new "UPDATE_REGISTRATION" to the table as well.

**Comment by chmielewskiandreas** [comment]
> I think this now looks as intended.
> 
> Can you also update the documentation and a diagram from here: https://docs.zephyrproject.org/latest/connectivity/networking/api/lwm2m.html#lwm2m-engine-and-application-events
> 
> The diagram is in a .png format that can be opened in [draw.io](https://app.diagrams.net/) then add the add the new "UPDATE_REGISTRATION" to the table as well.

Thx :) Due to refactoring I had to do minor refactoring (patch net: lwm2m: separate closing context from closing socket). I hope this approach is okay for you as well.

**Comment by SeppoTakalo** [comment]
@jheiskan81 can you review this also. I'm wondering if this clashes the same areas as your work on pause-resume topic.

**Review by jheiskan81 — DISMISSED** [review]
Yes true. https://github.com/zephyrproject-rtos/zephyr/pull/47327 Pause and stop functionality have a similar functionality:
Stop clear context and reset observation.
Pause only suspend connection and suspend thread.
Resume function validate similar way lifetime and enable thread: If still valid jump to update process otherwise to Full registration. There will be minor conflict but not a huge.

**Comment by SeppoTakalo** [comment]
build failures:
```
zephyr/subsys/net/lib/lwm2m/lwm2m_rd_client.c:190:3: error: implicit declaration of function 'lwm2m_socket_close' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                lwm2m_socket_close(client.ctx);
               
```

**Comment by chmielewskiandreas** [comment]
> build failures:
> 
> ```
> zephyr/subsys/net/lib/lwm2m/lwm2m_rd_client.c:190:3: error: implicit declaration of function 'lwm2m_socket_close' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
>                 lwm2m_socket_close(client.ctx);
>                
> ```

fixed :)

**Review by jheiskan81 — APPROVED** [review]

**Review by SeppoTakalo — APPROVED** [review]
Thank you for the thorough work you have done with this.
This is now looking clean.

**Comment by chmielewskiandreas** [comment]
> Thank you for the thorough work you have done with this. This is now looking clean.

thanks a lot for your reviews :)

**Review by rlubos — APPROVED** [review]
LGTM, thanks.

**Pull request merged**